hal/src/common/dmac/dma_controller.rs

```diff
@@ -0,0 +1,420 @@
+    #[inline(always)]
+    pub fn level_1_enabled(&mut self, enabled: bool) {
+        self.dmac.ctrl.modify(|_, w| w.lvlen1().bit(enabled));
+    }
+
+    /// Enable or disable priority level 2
+    #[inline(always)]
+    pub fn level_2_enabled(&mut self, enabled: bool) {
+        self.dmac.ctrl.modify(|_, w| w.lvlen2().bit(enabled));
+    }
+
+    /// Enable or disable priority level 3
+    #[inline(always)]
+    pub fn level_3_enabled(&mut self, enabled: bool) {
+        self.dmac.ctrl.modify(|_, w| w.lvlen3().bit(enabled));
+    }
+
+    /// Enable or disable Round-Robin Arbitration for priority level 0
+    #[inline(always)]
+    pub fn level_0_round_robin(&mut self, enabled: bool) {
+        self.dmac.prictrl0.modify(|_, w| w.rrlvlen0().bit(enabled));
+    }
+
+    /// Enable or disable Round-Robin Arbitration for priority level 1
+    #[inline(always)]
+    pub fn level_1_round_robin(&mut self, enabled: bool) {
+        self.dmac.prictrl0.modify(|_, w| w.rrlvlen1().bit(enabled));
+    }
+
+    /// Enable or disable Round-Robin Arbitration for priority level 2
+    #[inline(always)]
+    pub fn level_2_round_robin(&mut self, enabled: bool) {
+        self.dmac.prictrl0.modify(|_, w| w.rrlvlen2().bit(enabled));
+    }
+
+    /// Enable or disable Round-Robin Arbitration for priority level 3
+    #[inline(always)]
+    pub fn level_3_round_robin(&mut self, enabled: bool) {
+        self.dmac.prictrl0.modify(|_, w| w.rrlvlen3().bit(enabled));
+    }
+
+    /// Release the DMAC and return the register block
+    pub fn release(mut self, _pm: &mut PM) -> DMAC {
+        self.dmac.ctrl.modify(|_, w| w.dmaenable().clear_bit());
+
+        Self::swreset(&mut self.dmac);
+
+        #[cfg(any(feature = "samd11", feature = "samd21"))]
+        {
+            // Disable the DMAC clocking
+            _pm.apbbmask.modify(|_, w| w.dmac_().clear_bit());
+            _pm.ahbmask.modify(|_, w| w.dmac_().clear_bit());
+        }
+
+        // Release the DMAC
+        self.dmac
+    }
+
+    /// Issue a software reset to the DMAC and wait for reset to complete
+    #[inline(always)]
+    fn swreset(dmac: &mut DMAC) {
+        dmac.ctrl.modify(|_, w| w.swrst().set_bit());
+        while !dmac.ctrl.read().swrst().bit_is_clear() {}
+    }
+
+    /// Split the DMAC into individual channels
+    #[cfg(all(feature = "samd11", not(feature = "max-channels")))]
+    pub fn split(&mut self) -> DmacChannels {
+        DmacChannels(new_chan(), new_chan(), new_chan())
+    }
+
+    /// Split the DMAC into individual channels
+    /// Struct generating individual handles to each DMA channel
+    #[cfg(any(
+        any(all(feature = "samd11", feature = "max-channels")),
+        all(feature = "samd21", not(feature = "max-channels"))
+    ))]
+    pub fn split(&mut self) -> DmacChannels {
+        DmacChannels(
+            new_chan(),
+            new_chan(),
+            new_chan(),
+            new_chan(),
+            new_chan(),
+            new_chan(),
```

Comment:
Consider using the `seq_macro` crate to generate the `split` function and `DmacChannels` struct. It's basically a for loop in a macro. To do so, you might need to define your `NUM_CHANNELS` constant in a macro, but that's an easy change. Check the closed issues on the `seq_macro` GitHub for a clearer explanation of that last point.